app/models/applicant.rb

```diff
@@ -7,12 +7,15 @@ class Applicant < ApplicationRecord
   has_many :log_entries
   belongs_to :campus
 
-  attr_accessor :password, :password_confirmation, :old_password, :gdpr_checkbox
+  attr_accessor :email_confirmation, :password, :password_confirmation, :old_password, :gdpr_checkbox
 
   validates :firstname, :surname, :email, :phone, :campus, presence: true
   validates :email, :phone, uniqueness: true
 
-  validates :email, email: true
+  validates :email, :email_confirmation,
+            presence: { if: ->(applicant) { applicant.new_record? } }
+  validates :email, confirmation: { if: ->(applicant) { applicant.new_record? } }
+  validates :email, format: { with: /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i }
```

Comment:
This regex excludes some special characters that are allowed in emails, it seems. !#$%&'*+-/=?^_`{|}~ are all legal characters in the first part of the email( some more info here: https://en.wikipedia.org/wiki/Email_address#Syntax).

Reply:
This means that the start of the email can contain these characters?
I am not well versed in regexp, do you have any suggestions for how to improve it?

Reply:
This `^(?:[a-z0-9!#$%&'*+\/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+\/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])$` from [here](https://stackoverflow.com/questions/201323/how-can-i-validate-an-email-address-using-a-regular-expression) seems to catch more(but not all, like " "@example.com or "very.(),:;<>[]\".VERY.\"very@\\ \"very\".unusual"@strange.example.com).

But I also think there is a gem for email validation, but I haven't tested it: https://github.com/K-and-R/email_validator?ref=magicbell.ghost.io

Reply:
Personally, I think that seems a bit overkill.
We allow having + . _ and - in the first part, and I think that is more than enough. No one uses emails containing those symbols either way :^)

Reply:
I agree, but as soon as we push this verification, the one person in the world with an email containing a dollar sign will apply. This regex is less strict(allowing more special characters):
**\A[\w+\-.!#$%&'*\/=?^_`{|}~]+@[a-z\d\-.]+\.[a-z]+\z/i**
and this one allows everything(as long as it ends with and @ and a domain name):
**\A.+@[a-z\d\-.]+\.[a-z]+\z/i**
This one ensures only one @ (even though more are allowed if contained in quotation marks):
**\A[^@]+@[a-z\d\-.]+\.[a-z]+\z/i**

Reply:
> I agree, but as soon as we push this verification, the one person in the world with an email containing a dollar sign will apply. This regex is less strict(allowing more special characters): **\A[\w+-.!#$%&'*/=?^_`{|}~]+@[a-z\d-.]+.[a-z]+\z/i** and this one allows everything(as long as it ends with and @ and a domain name): **\A.+@[a-z\d-.]+.[a-z]+\z/i** This one ensures only one @ (even though more are allowed if contained in quotation marks): **\A[^@]+@[a-z\d-.]+.[a-z]+\z/i**

Changed the regex slightly, according to this article:
https://www.abstractapi.com/guides/ruby-email-regex

I've tried adding some of your tricky regexp rules in ruby, but they end up being rendered as comments (it's not as easy as just copy pasting it into there). This at least allows one special character, as long as it is followed by a letter afterwards.